### PR TITLE
[4/n] Thread PG: add reduce_scatter to threaded pg

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89443
* __->__ #89442
* #89441
* #89440
* #89439

### CODE REVIEW

**Review by yhcharles — APPROVED** [review]
LGTM

File: torch/testing/_internal/distributed/multi_threaded_pg.py [hunk 1]
```diff
@@ -69,6 +70,30 @@ def work(self, data):
...
+                dest_tensor_on_rank_i = data[i][0]
+                # Can't handle reduce_scatter with multiple output tensor
+                assert len(dest_tensor_on_rank_i) == 1
+                if not start_reduction[i]:
+                    with torch.no_grad():
+                        dest_tensor_on_rank_i[0].copy_(to_scatter[i])
+                    start_reduction[i] = True
+                else:
+                    with torch.no_grad():
+                        dest_tensor_on_rank_i[0].add_(to_scatter[i])
```
Comment (fduwjj) [inline torch/testing/_internal/distributed/multi_threaded_pg.py:95]: Interesting, does this mean the overall aggregation logic is executed in one single thread?

**Review by fduwjj — APPROVED** [review]
LGTM, just have a n00b question.

File: torch/testing/_internal/distributed/multi_threaded_pg.py (hunk 1 above)
Comment (wanchaol) [reply torch/testing/_internal/distributed/multi_threaded_pg.py:95]: yep, it's not actually running in multi threads, only PG is mocked with multiple threads to assume it's a world.

**Comment by wanchaol** [comment]
@pytorcbot merge -f "failures not related"

**Pull request closed without merging**